### PR TITLE
[MLIR] Add patterns to bubble-up pack and push-down unpack through collapse/expand shape ops

### PR DESCRIPTION
**Pull request by pzread**
Add DataLayoutPropagation patterns to bubble-up pack and push-down unpack through collapse/expand shape ops.

This is possible when the inner tile sizes of pack/unpack can divide the projected dims after being swapped with collapse/expand shape ops. For example:

```mlir
%collapsed = tensor.collapse_shape %1 [[0], [1, 2]] : tensor<3072x64x4xf32> into tensor<3072x256xf32>
%pack = tensor.pack %collapsed outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 1] into %2 : tensor<3072x256xf32> -> tensor<384x256x8x1xf32>
```

can be transformed into:

```mlir
%pack = tensor.pack %1 outer_dims_perm = [0, 1, 2] inner_dims_pos = [0, 2] inner_tiles = [8, 1] into %2 : tensor<3072x64x4xf32> -> tensor<384x64x4x8x1xf32>
%collapsed = tensor.collapse_shape %pack [[0], [1, 2], [3], [4]] : tensor<384x64x4x8x1xf32> into tensor<384x256x8x1xf32>
```

The current support cases are:
- Bubble-up pack op through `tensor.collapse_shape`
- Push-down unpack op through `tensor.expand_shape`

### CODE REVIEW

**Comment by pzread** [comment]
Kindly ping : )

File: mlir/test/Dialect/Linalg/data-layout-propagation.mlir [hunk 1]
```diff
@@ -905,3 +905,127 @@ func.func @unpack_different_destination_shape(%arg0: tensor<1x1x1080x1920x16xi32
...
+  %6 = tensor.empty() : tensor<3072x256xf32>
+  %unpack = tensor.unpack %5 outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 8] into %6 : tensor<384x32x8x8xf32> -> tensor<3072x256xf32>
+  %expanded = tensor.expand_shape %unpack [[0, 1], [2]] : tensor<3072x256xf32> into tensor<256x12x256xf32>
+  func.return %expanded : tensor<256x12x256xf32>
+}
+// CHECK-LABEL: func.func @no_push_down_unpack_through_non_divisible_expand
+// CHECK-SAME:      %[[ARG0:[a-zA-Z0-9]+]]
+// CHECK:         %[[UNPACK:.+]] = tensor.unpack %[[ARG0]]
+// CHECK:         %[[EXPANDED:.+]] = tensor.expand_shape %[[UNPACK]] {{\[}}[0, 1], [2]] : tensor<3072x256xf32> into tensor<256x12x256xf32>
+// CHECK:         return %[[EXPANDED]] : tensor<256x12x256xf32>
```
Comment (qedawkins) [inline mlir/test/Dialect/Linalg/data-layout-propagation.mlir:1031]: Can you add a test where the reassociated dims are not affected by the pack (or unpack)? e.g.
```mlir
  %collapsed = tensor.collapse_shape %1 [[0, 1], [2]] : tensor<?x16x4xf32> into tensor<?x4xf32>
  %2 = tensor.empty(%dim) : tensor<?x2x2xf32>
  %pack = tensor.pack %collapsed inner_dims_pos = [1] inner_tiles = [2] into %2 : tensor<?x4xf32> -> tensor<?x2x2xf32>
```

**Review by qedawkins — APPROVED** [review]
Sorry for the late review, LGTM, thanks!

File: mlir/test/Dialect/Linalg/data-layout-propagation.mlir (hunk 1 above)
Comment (pzread) [reply mlir/test/Dialect/Linalg/data-layout-propagation.mlir:1067]: Done.

**Pull request merged**